src/Discord/Internal/Types/Interactions.hs

```diff
@@ -0,0 +1,363 @@
+              ("name", toMaybeJSON interactionDataApplicationCommandName),
+              ("type", toMaybeJSON interactionDataApplicationCommandType),
+              ("resolved", toJSON <$> interactionDataResolved),
+              ("options", toJSON <$> interactionDataOptions),
+              -- missing info relevant for components
+              --   , ("custom_id", toJSON <$> interactionDataCustomId)
+              --   , ("component_type", toJSON <$> interactionDataComponentType)
+              ("target_id", toJSON <$> interactionDataTargetId)
+            ]
+      ]
+
+instance FromJSON InteractionData where
+  parseJSON =
+    withObject
+      "InteractionData"
+      ( \v ->
+          InteractionData
+            <$> v .: "id"
+            <*> v .: "name"
+            <*> v .: "type"
+            <*> v .:? "resolved"
+            <*> v .:? "options"
+            -- <*> v .:? "custom_id"
+            -- <*> v .:? "component_type"
+            -- <*> v .:? "values"
+            <*> v .:? "target_id"
+      )
+
+-- | I'm not sure what this stuff is, so you're on your own.
+--
+-- It's not worth the time working out how to create this stuff.
+-- If you need to extract from these values, check out the link below.
+--
+-- https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-object-resolved-data-structure
+data ResolvedData = ResolvedData
+  { resolvedDataUsers :: Maybe Value,
+    resolvedDataMembers :: Maybe Value,
+    resolvedDataRoles :: Maybe Value,
+    resolvedDataChannels :: Maybe Value
+  }
+  deriving (Show, Read, Eq)
+
+instance ToJSON ResolvedData where
+  toJSON ResolvedData {..} =
+    object
+      [ (name, value)
+        | (name, Just value) <-
+            [ ("users", resolvedDataUsers),
+              ("members", resolvedDataMembers),
+              ("roles", resolvedDataRoles),
+              ("channels", resolvedDataChannels)
+            ]
+      ]
+
+instance FromJSON ResolvedData where
+  parseJSON =
+    withObject
+      "ResolvedData"
+      ( \v ->
+          ResolvedData
+            <$> v .:? "users"
+            <*> v .:? "members"
+            <*> v .:? "roles"
+            <*> v .:? "channels"
+      )
+
+-- | The application command payload for an interaction.
+data ApplicationCommandInteractionDataOption = ApplicationCommandInteractionDataOption
+  { applicationCommandInteractionDataOptionName :: String,
+    applicationCommandInteractionDataOptionType :: ApplicationCommandOptionType,
+    -- | The value itself. Mutually exclusive with options.
+    applicationCommandInteractionDataOptionValue :: Maybe StringNumber,
+    -- | Only present in group subcommands and subcommands. Mutually exclusive with value.
+    applicationCommandInteractionDataOptionOptions :: Maybe [ApplicationCommandInteractionDataOption],
+    -- | Whether this is the field that the user is currently typing in.
+    applicationCommandInteractionDataOptionFocused :: Maybe Bool
+  }
```

Comment:
:+1:  for the names. Tough to put "mutually exclusive with value" in a comment. Ideally this is abstracted into an ADT to express this, but I get that this is all WIP on new docs. Not worth it now.

Reply:
My plan is for _all_ of this stuff to be made "Internal", and then create ADTs that translate into them, so the internals are replicas of the input JSON and the ones that we advertise to users are usable in haskell.

Reply:
That's a good idea